### DISCUSSION TITLE
Gear Sensor

### DISCUSSION THREAD

**Discussion by polytruand** [danielnilsson9/bbs-fw#67] [Ideas]
Hello Daniel Nilsson, First of all, thank you very much for the work done, your firmware is great for the configuration.
However, I have a question:
- Does it support shift sensor? If so, is it possible to change the settings? 

Thank you for your help and wish you a pleasant weekend.

**Comment by Zensaition** [comment]
for my bbso2 I noticed its not working or the firmware knows its there.

**Comment by danielnilsson9** [comment]
Yes, shift sensor support is implemented and working. However I have no idea of how well it works compared to the original firmware. I don't have a gear sensor myself so I'm not able to test anything there. The current setting was selected by a uses which did some tests. The interrupt i 600ms using a fast power ramp down to 50% of target current.

You cannot change anything in settings. I will not do any development on shift sensor stuff since I'm unable to test it. However if anyone want to improve the current situation, please feel free to do so and open a pull request to contribute the code.

**Comment by CitizenDev** [comment]
From my testing it definitely does something, Easiest method is to hold throttle and then change gears. But I agree that it's not yet up to the same quality as original firmware.
I selected the longer 600ms ramp down but it either does not ramp down fast enough or 50% target current is still too high as it can still be almost 700W is running at max current at time of shift. I think we should try to set it around 250W max at the time of the shift but since I haven't been driving much because of bad weather I haven't tried it out yet. Daniel did show me where the current ramp is calculated in the code in order to do the changes for testing.

Logic is defined in app.c file starting at row 672 function apply_shift_sensor_interrupt if you want to have a go yourself.

**Comment by dav0000000** [comment]
I think the shift sensor is a bit of a personal requirement based heavily on the parts in one's bike. My wife's bike has nice sharp gear changes but mine are terrible. I took the code quite a while ago when Daniel was no longer supporting changes and got it working really good. In my version I do instant ramp down to 0 for gear change, then I maintain 0 for 120ms. Then I ramp from 1 to 20 target_current in 1 current increments at 20ms duration then go full target-current. If the speed is under 10km/h I ramp up from 1 to 20 in 100ms increments. I will try to feed this back into the project but I'm doing various parts a lot differently and in the end I've done this primarily for my crappy drivetrain. I can configure these values by making changes to an array so it starts to get confusing for the average user. I have different ramp up and down speeds for example a pas change which is very fast at 5 target-current every 50ms to slow ramp up at 1 every 100ms. I do fast ramp up at 5 target_current per 50ms over 18km/h. I ramp down instantly under 8km/h. This all works great for me but putting my changes into something for all users makes it challenging. This is the beauty of open source and sometimes you just need to do some programming yourself. I'm still testing all this and I feel it works great but I need a little more time to feed something useful back into the project.

**Reply by zeynelozturk** [reply]
If more information will be useful in future, here is my experience:  Shifting (especially to harder gear) a bit harsh shifts under power >200W on my fat bike. It's fine at about 80W.  

I'm definitely not expert in this but if the delay/ramp down would be relative to current motor power, maybe it'd be better (than relative to current bike speed)? Or perhaps as dav0000000 said, making it relative to speed is better, making the pause shorter for climbing hills...

**Reply by CitizenDev** [reply]
Hey @dav0000000 Would you care sharing your code?
I might try to derive something from your code that could be implemented in public release.

**Reply by dav0000000** [reply]
My thought was that the gear change ramp up power should be consistent every time and not based on power level. In relation to the slow gear change ramp up under 10km/h, If I am changing gears at this speed (without the brakes cutting the motor) then I have stuffed up and stopped in the wrong gear. At slow speeds, to much power results in crunching due to the pedals turning too slowly and the gears not changing quick enough. I don't think my settings are right for everyone but I think perhaps a ramp up based on a configurable current level would work best using Daniel's existing code. So instead of using target_current for the calculation just use a set value unless target_current is lower than the set value.

**Comment by BotoX** [comment]
Hi, how about this idea:
Cut the power (or reduce to max ~50 Watt) for at least one whole wheel revolution?
The speed sensor measures wheel revolutions and the cassette is connected directly to the wheel, the chain needs around one revolution average to fully change gears. Or just make it configurable.
The implementation shouldn't depend on the actual raw wheel revolutions (counts) but use the delay between the last two counts (or whatever method is used to calculate the speed in km/h without applying the wheel diameter)

**Reply by cdbrow** [reply]
Hello, this has been an issue for me as well. I’ve gotten to the point where I just stop peddling while shifting. Daniel shared with us where we can make the change in the code. It took me a while to figure out because I don’t really do coding. His firmware is set to stop power for 600 milliseconds, I have updated the firmware to stop power for 1200 milliseconds. I just have not had the opportunity to upload the firmware for testing.  I will try to do that today. If it works well, I could share out the image.

**Reply by wing-dinger** [reply]
I wouldn't be able to make this change without guidance, but also need a longer pause with lower power in order to smooth shifting. If you make progress I'd love to try a change.

**Comment by cdbrow** [comment]
I updated the following lines of code. 

Line 71 from 600 to 1200 (this is the total delay for power being cut during shifting)
// How long the power interrupt will last when gear sensor is triggered.
#define SHIFT_SENSOR_INTERRUPT_PERIOD_MS		1200

Lines 692, 697, and 702 (before it was reducing the power to 3/4 at beginning and end of shift, it is now set to 1/4, before it was reducing power to 1/2 during the shift and is now set to 1/6)
		if (timediff < SHIFT_SENSOR_INTERRUPT_PERIOD_MS / 4)
		{
			// reduce target power to 1/4 of requested (ramp down), shift started
			*target_current = (uint8_t)(*target_current / 4u);
		}
		else if (!active && timediff > (3 * SHIFT_SENSOR_INTERRUPT_PERIOD_MS) / 4)
		{
			// reduce target power to 1/4 of requested (ramp up), shift finished
			*target_current = (uint8_t)(*target_current / 4u);
		}
		else
		{
			// keep power at 1/6 requested
			*target_current = (uint8_t)(*target_current / 6u);

Here is a link to download the updated FW, based on the above changes. Let me know how it works out for you. I would like to set it to a target current of like 200 Watts throughout the process, but I don't know how to specify that.
https://1drv.ms/u/s!AmjEnig9JYIF-t4AiwOYzKHPi3Tp5A?e=fZXNzN

**Reply by wing-dinger** [reply]
Awesome! My bike is in pieces for another day or two, I'll try to test this change before the end of the weekend. Thanks!

**Reply by cdbrow** [reply]
I tried it last night and I felt like the delay was too much, so I've changed it back to 600ms, I also updated the power ramp down a bit to the following. 

I'm having it reduce power down to 1/8 at the beginning and middle of the shift, 1/4th at the end of the shift. 

Here is the link to download this version of the FW.
https://1drv.ms/u/s!AmjEnig9JYIF-t4cgRuXTHvmDpBCSw?e=HOJ6lb

Also, I learned when trying to install this version that it required a newer Config Tool, that is because I had grabbed the latest branch when I started working with it and Daniel had already added code for Keep Current and Keep Current Cadence. So I had to compile the code for his version 1.3.0 of the Config Tool.  Here is the link to download that. You will need to extract the zip file and then run BBSFWTool.exe when ready. 

https://1drv.ms/u/s!AmjEnig9JYIF-t4UR5mlyY56rrc-Eg?e=KldIGz

Keep in mind that upgrading the FW will set your config back to default. You should use the old Config Tool to backup your config, then upgrade your FW, then open the new Config Tool to import and restore your backup config. Also note that when doing this, you need to set the values for Keep Current and Keep Current Cadence, otherwise you can't write the config, it will error out. The defaults are Keep Current 60 and Keep Current Cadence 40.

**Reply by cdbrow** [reply]
No problem! Without really thinking, I had downloaded his latest code, which has not been compiled and released yet. That code included the firmware, and the config tool. I used Make to compile the firmware after making the changes. I use visual studio to compile the config tool. I’m really noob with coding, I know I didn’t do the config tool properly because the exe doesn’t contain everything, which is why I zipped up all of the files, then just run the exe from the folder that contains the exe. The first time I tried to run the tool and write the config changes, it actually crashed, but worked the second time. Keep in mind that it’s basically better because Daniel hasn’t released it for one reason or another yet.

**Reply by cdbrow** [reply]
I should probably modify the 1.2.0 code so that the released version of the config tool will work. It just won’t contain the modification that Daniel has added for keep current. 

If anyone wants me to do this, I can, just let me know.

**Reply by wing-dinger** [reply]
I just flashed this new version, and did a short test ride. The shifts feel good! It's a bit more like a manual transmission, with a noticeable pause in power during the shift, then a ramp back up after the new gear engages. With decently tuned shifting the pause is long enough and power drop seems sufficient (though I didn't do any very high power shifts, I just try to avoid them in general). I'll update if this impression changes after more riding.

Keep current question: What settings (Keep Current and Keep Current Cadence) will mimic the V1.2.0 firmware? Is it 100/100 (no power reduction with increased cadence), 60/40 (the stated defaults), or something else? I have not played around with this setting much but will do so and see what behavior I like most.

Thanks!!

**Reply by cdbrow** [reply]
I’m glad you got a chance to try it. Sounds like we may need to reduce the delay a bit more, lower than 600ms. To be honest I still need to try it. 

Regarding the keep current settings, I’m not completely sure. I think setting the keep current to 100 would mimic 1.2.0. The cadence setting would only kick in if keep current was something less than 100.

**Reply by wing-dinger** [reply]
On my heavy cargo bike, the delay of 600ms is good. Shorter will probably cause the power to ramp back up during a slow shift. 
I just tried keep current/cadence at 60/60 and I think I can feel a power reduction as I pedal faster? But I probably wouldn't have noticed a difference if I didn't know. Will play with it more later.

**Reply by EigenMania** [reply]
@cdbrow That's exactly right. If you look at line 357 of app.c (in the firmware source code), you'll see that if the keep current percent is set exactly to 100, then the entire block will be skipped.

Also, dropping the power to a fixed level is possible. You would just take the desired power level during shift and compute the necessary target current percent based on the max allowed current and battery voltage. Then of course you should care only to apply this if the target current (pre-shift) was only higher than this threshold (otherwise for low power pedalling it would power up during the shift).

What if the config tool had a section for shifting? Then the shift pause (ms) and shift power drop (W) could be tuned from the GUI rather than the FW. If you think this makes sense, I could give it a shot and share the results here.

**Reply by wing-dinger** [reply]
If the config tool allowed tuning of shift power I would use it! Up to you how much effort you want to put in. I'm pretty happy with the current behavior.

**Reply by cdbrow** [reply]
@EigenMania I would definitely like the ability to make changes to the shift delay/power via the tool, rather than having to update the FW. Some people will have different configs and want it to be different.

**Comment by EigenMania** [comment]
Here is my attempt :)

I modified the firmware and config tool to allow for configurable shifting behaviour. You can adjust the shift pause time (ms) and the maximum power allowed through the drivetrain (W) during a shift. So if you are below this power then nothing will happen, and above there will be a power cut down to the threshold power. Tested on my BBSHD and so far seems to work ;)

Firmware: https://www.dropbox.com/s/v155vtyjqm4mwjk/2022-12-03_fw_tuneable_shift_sensor.hex?dl=0
Config: https://www.dropbox.com/s/lci62c1dngmz7z2/2022-12-03_fw_tool_tuneable_shift_sensor.exe?dl=0

Or grab from my branch: https://github.com/EigenMania/bbs-fw/tree/feature/tuneable_shift_sensor

Feedback appreciated!

**Reply by cdbrow** [reply]
Thank you! I will definitely check out your branch. Did you modify the code from his latest version that also has keep current? It sounds like you have a lot more experience with programming than I do, so I appreciate you jumping in. I was really just fumbling around and playing, but was grateful that it did work.

**Reply by EigenMania** [reply]
Yes, it's based on the most recent code with keep current (which you can disable by setting Keep Current (%) = 100 if desired.

**Reply by kklmm** [reply]
> Yes, it's based on the most recent code with keep current (which you can disable by setting Keep Current (%) = 100 if desired.

@EigenMania would you please send a pull request to daniel so your code can be merged into his master branch? I see you already sent him a pull for a variable fix. it'd be totally awesome if we can have your changes in his master repo. daniel's firmware is excellent but lack of power cut off during shifting makes it a deal breaker for some of us. I broke two teeth on 4th gear two weeks ago with his firmware and immediately went back to stock. I'd love to go back but not until power properly cuts off during shift like stock firmware does. Using just your branch isnt an option either as I'm locking myself out of future updates and fixes from daniel's master.

**Reply by EigenMania** [reply]
@kklemen I will submit a PR eventually, I just need to test a bit more. On my current build I don't have the shift sensor integrated, so I just plugged it in and triggered it manually with a piece of cable while riding. Nothing caught fire (and I didn't crash my bike) so I reckon the changes are pretty safe ;) I have a BBS02 build back home with a shift sensor properly integrated. So I will do some high power shifts on full throttle to make sure it really works as intended. Just has to wait until I'm back for the holidays. Then I can also suggest some default values and reasonable ranges for the parameters on the GUI.

**Reply by danielnilsson9** [reply]
Thank you for helping out! Just a heads up here. I looked at you implementation quickly and as far as getting this merged there are a few problems thats needs to be addressed. There are multiple float calculations which are very expensive on a 8bit mcu without an fpu. If you look through the rest of the code there are mostly fixed point calculations done (apart from temperature sensor calculations which should probably be fixed at some point).

I do see the logic of configuring this in watts but I have specifically avoided that in order to minimize the amount of calculations needed to be done every loop. All other power related stuff is configured in percent of maximum power.

This might sound silly since you managed to build it, flash it any run it but we have to think about the long run and avoid unnecessary computation.

I'm currently porting this firmware to TSDZ2 for my other bike, the TSDZ2 has an 8bit MCU running at 16MHz which is also responsible for the motor control logic. This means there are very little cpu resources left over for running the rest of the firmware... I don't yet know if it will be able to run this firmware without more optimizations.

I hope I don't come of as rude, I just wanted to let you know.

**Reply by EigenMania** [reply]
Not rude at all, you definitely have more experience than me and that is all valuable input. I'm maybe a bit spoiled by all of these fancy modern microcontrollers that now run in GHz ranges and you can just do whatever the hell you want in Python. Maybe one day we will also have a BBSHD that can do machine learning :P 

Shouldn't be too hard to do this with respect to percent current / power, and use the target values (when the shift starts) instead of the actual measurements to minimize computation. Maybe the user could also provide their nominal battery voltage (ie 36, 48, 52, ...) in the GUI so that the power threshold estimates are more accurate.

**Reply by kklmm** [reply]
@EigenMania flashed your hex. tried using your tool from dropbox dl link. no go.

flash -> read -> failed to read configuration from flash. timeout occurred. flash -> reset -> same issue. cannot see any inputs for adjusting shifting behaviour either. are you sure this is a proper binary of config tool in the link above?

**Reply by EigenMania** [reply]
@kklemen Ahh shoot sorry, that's embarassing. Maybe I goofed and uploaded the wrong Config tool... Would explain the config reading timeout since the firmware and config probably don't match. 

Here are updated compiles: 
Firmware: https://www.dropbox.com/s/uzr9jlkqbdxkcld/2022-12-15_fw_tuneable_shift_sensor.hex?dl=0
Tool: https://www.dropbox.com/s/sv85tkmk658qujt/2022-12-15_fw_tool_tuneable_shift_sensor.exe?dl=0

They should have the bugfix we merged recently (for the speedo cutoff) and some computation optimizations that Daniel suggested. 

Please update me if there are still issues on your end :)

**Reply by kklmm** [reply]
Thank you for rapid reply!

unfortunately no change. same issue. flashed the firmware and used your  latest binary 2022 12 15.
Please correct me if im using wrong approach. I place your .exe inside daniel's config tool folder containing his dll's and run your exe instead of his, yes? Otherwise without his dll's your binary does not run.

**Reply by EigenMania** [reply]
Oh I didn't realize that the .exe doesn't run without the other files. I've always launched it from the build folder (locally). 

Here is the entire folder (zipped): https://www.dropbox.com/s/tsv16w6mx3g4azt/2022-12-15_fw_tool_tuneable_shift_sensor.zip?dl=0

**Reply by kklmm** [reply]
worked like a charm. thank you for the link. drove around for 2 hours nearly depleting entire 52v 20ah battery. bike rides like a dream. shifting is smooth. acceleration graceful yet plentiful. absolute joy. smiled the entire way despite the ice and freezing winter temperatures. thanks a bunch! flashing this next onto all of the other bafang engines.

**Reply by manusalgam** [reply]
I supposed this .hex is for BBSHD, so please, may you build this fw to BBS02B (or share the code, the branch appears to be nonexistent)? Thanks a lot!

**Reply by kb79000** [reply]
> Je suppose que ce .hex est pour BBSHD, alors s'il vous plaît, pouvez-vous construire ce fw sur BBS02B (ou partager le code, la branche semble inexistante) ? Merci beaucoup!

Hello, Did you manage to operate the sensor on your BBS02B?

**Comment by dav0000000** [comment]
Hi guys, I've been following this discussion closely. I did have the shifting working pretty good in my own version. I've described my approach before in this thread. I was never completely happy though and could never get daniel's approach working well. When @BotoX mentioned the wheel revolution idea I ran with it immediately. It was a great idea and made things much better across different speeds. I settled on 0% current during the shift quite a while ago. I've spent so much time testing different shift currents and it would be so smooth and fast one second and absolutely terrible on the drivetrain the next. After the wheel speed change I did try again but was very unhappy with it. So 0% seems to be the best for me and my drivetrain. Now, I had one problem left which was the chain would jump around a bit on shift. I would watch my chain wobble while shifting and sometimes fall off and had an idea to increase the current very quickly when shift was detected. I thought it would crunch but it worked. Now I'm happy with the shifting for the first time ever. I made an update in my fork to show the way I run the shift sensor currently. The code is not the best as unfortunately I run my own version and handle the gear changes in apply_ramp_up and apply_ramp_down so it was a little tricky and I had to do ramp up within the shifting function. It works though. I would love to hear any feedback you may have. If you have any questions please feel free to ask.

**Reply by zeynelozturk** [reply]
I'm hoping a nice solution like this will be merged into bbs-fw in near future. I'm using this firmware but earlier with also original firmware, I noticed I always use "ebrake cutoff" method when changing multiple gears (currently three at most), to avoid crunching.

Does this solution handles changing multiple gears better? (I don't know if gear sensor signal allows to distinguish it).

**Reply by EigenMania** [reply]
I'll definitely try it out! I looked at the code and left a few comments in the commits. I especially like the idea to ramp back up to target_current after the shift, but I think this could be done a bit more concisely with the MAP32 function. I also like the idea of a shift pause that is speed dependent. But rather than doing it as a function of wheel speed, I think it makes more sense to do it as a function of pedal cadence (when the shift is initiated) because it is pedalling action that completes the shift (and not rolling of the rear wheel).

**Reply by kklmm** [reply]
@dav0000000 thanks for sharing your code. can you provide compiled version of .exe and .hex firmware? thanks

**Comment by dav0000000** [comment]
It doesn't matter how fast or slow you are pedalling, we just have to get that chain wrapped around the rear cog. The pedal cadence is not real useful as the motor can easily have a different speed to the pedals. In my code (not this code) I ramp up 2x faster than this exponential ramp when over 25km/h and 3x faster when over 35km/h. The shifting at speed is not the issue for me, it was always at the lower speeds. Try it anyway.

**Reply by dav0000000** [reply]
> In my code (not this code) I ramp up 2x faster than this exponential ramp when over 25km/h and 3x faster when over 35km/h.

I checked my code and this is not correct. I am only doing 1x on gear ramp. I tested against my version and it is a little different on ramp. It is not quite as fast. My personal code ramps up every 25ms whereas the default ramp up in this code is 20ms. I am going to add a define and set the ramp_up to 25ms. Just hoping someone can take the time to test this and tell me how it compares. A question I have is whether @danielnilsson9 got the code he is using from the original firmware or not?

**Reply by danielnilsson9** [reply]
No, no code is from the original firmware. The current speed sensor implementation was just some simple stuff I made without any ability to test it... So don't worry about the current implementation, its bad. I never investigated how the interrupt behaved in the original firmware.

**Comment by kklmm** [comment]
Should be the same file. Bbsfw is supposed to be universally compatible.
Only settings are different.

**Reply by manusalgam** [reply]
Before compile the .hex, prompt asks me for the target device (BBSHD, BBS02B, and another one which doesn't matter now), and both devices has different controllers (with different amount of ROM and RAM, MOSFETs, different IO pins, etc.). Before flashing I'd rather to be truly sure, it could be dangerous or may brick my motor.

**Comment by kb79000** [comment]
Hello, today I installed version 1.3.0 on my BBS02B I noticed a fault with the shifter I tried several settings but I have the impression that the power cut is not happening have you also noticed  this issue ?

**Reply by skeeterman007** [reply]
I wrote a version of the firmware which allows you to turn of the target current ramp down in the gui, as the target current ramp down on the latest version interferes with the shift sensor and wont let power be cut quick enough. its called smooth shift and its enable able in the tool, also it works very well with the pretension feature.
https://github.com/skeeterman007/bbs-fw/tree/Smooth_Shift

**Comment by CitizenDev** [comment]
> I wrote a version of the firmware which allows you to turn of the target current ramp down in the gui, as the target current ramp down on the latest version interferes with the shift sensor and wont let power be cut quick enough. its called smooth shift and its enable able in the tool, also it works very well with the pretension feature. https://github.com/skeeterman007/bbs-fw/tree/Smooth_Shift

Does this fork have pretension feature in it?
